Fails to build on FreeBSD PowerPC64

Hi, sysutils/lsof fails to build on FreeBSD / PowerPC64 with the following error:

cc -o lsof  -pipe -fstack-protector-strong -fno-strict-aliasing -DNEEDS_BOOL_TYPEDEF -DHASTASKS -DHAS_PAUSE_SBT -DHAS_DUP2 -DHAS_CLOSEFROM -DHASEFFNLINK=i_effnlink -DHASF_VNODE -DHAS_FILEDESCENT -DHAS_TMPFS -DHASWCTYPE_H -DHASSBSTATE -DHAS_KVM_VNODE -DHAS_UFS1_2 -DHAS_NO_IDEV -DHAS_VM_MEMATTR_T -DNEEDS_DEVICE_T -DHAS_CDEV2PRIV -DHAS_NO_SI_UDEV -DHAS_SYS_SX_H -DHASFUSEFS -DHAS_ZFS -DHAS_V_LOCKF -DHAS_LOCKF_ENTRY -DHAS_NO_6PORT -DHAS_NO_6PPCB -DNEEDS_BOOLEAN_T -DHAS_SB_CCC -DHAS_FDESCENTTBL -DFREEBSDV=13000 -DHASFDESCFS=2 -DHASPSEUDOFS -DHASNULLFS -DHASIPv6 -DHASUTMPX -DHAS_STRFTIME -DLSOF_VSTR=\"13.0-CURRENT\" -I/usr/src/sys -O2 dmnt.o dnode.o dnode1.o dnode2.o dproc.o dsock.o dstore.o  arg.o main.o misc.o node.o print.o proc.o store.o usage.o  util.o -L./lib -llsof  -lkvm
dnode2.o:(.text+0x190): undefined reference to `abd_zero_off'
dnode2.o:(.text+0x1c8): undefined reference to `abd_cmp_buf_off'
dnode2.o:(.text+0x200): undefined reference to `abd_copy_to_buf_off'
dnode2.o:(.text+0x238): undefined reference to `abd_copy_from_buf_off'
dnode2.o:(.text+0x274): undefined reference to `abd_copy_off'
*** Error code 1

Stop.
make[1]: stopped in /wrkdirs/usr/ports/sysutils/lsof/work/lsof-4.93.2
*** Error code 1

Stop.
make: stopped in /usr/ports/sysutils/lsof

This is from: http://pylon.nyi.freebsd.org/data/head-powerpc64-default/p506720_s350020/logs/errors/lsof-4.93.2_2,8.log

There is a mention of some one getting it to pass by adding -lzpool in the email posting: http://freebsd.1045724.x6.nabble.com/lsof-doesn-t-compile-on-my-ppc64-td6294887.html

Regards

Hi, I'm currently getting another error, I'm running FreeBSD Current:

(cd lib; make DEBUG="-g" CFGF="-pipe -DNEEDS_BOOL_TYPEDEF -DHASTASKS -DHAS_PAUSE_SBT -DHAS_DUP2 -DHAS_CLOSEFROM -DHASEFFNLINK=i_effnlink -DHASF_VNODE -DHAS_FILEDESCENT -DHAS_TMPFS -DHASWCTYPE_H -DHASSBSTATE -DHAS_KVM_VNODE -DHAS_UFS1_2 -DHAS_NO_IDEV -DHAS_VM_MEMATTR_T -DNEEDS_DEVICE_T -DHAS_CDEV2PRIV -DHAS_NO_SI_UDEV -DHAS_SYS_SX_H -DHASFUSEFS -DHAS_ZFS -DHAS_V_LOCKF -DHAS_LOCKF_ENTRY -DHAS_NO_6PORT -DHAS_NO_6PPCB -DNEEDS_BOOLEAN_T -DHAS_SB_CCC -DHAS_FDESCENTTBL -DFREEBSDV=13000 -DHASFDESCFS=2 -DHASPSEUDOFS -DHASNULLFS -DHASIPv6 -DHASUTMPX -DHAS_STRFTIME -DLSOF_VSTR=\"13.0-CURRENT\"")
cc   -pipe -DNEEDS_BOOL_TYPEDEF -DHASTASKS -DHAS_PAUSE_SBT -DHAS_DUP2 -DHAS_CLOSEFROM -DHASEFFNLINK=i_effnlink -DHASF_VNODE -DHAS_FILEDESCENT -DHAS_TMPFS -DHASWCTYPE_H -DHASSBSTATE -DHAS_KVM_VNODE -DHAS_UFS1_2 -DHAS_NO_IDEV -DHAS_VM_MEMATTR_T -DNEEDS_DEVICE_T -DHAS_CDEV2PRIV -DHAS_NO_SI_UDEV -DHAS_SYS_SX_H -DHASFUSEFS -DHAS_ZFS -DHAS_V_LOCKF -DHAS_LOCKF_ENTRY -DHAS_NO_6PORT -DHAS_NO_6PPCB -DNEEDS_BOOLEAN_T -DHAS_SB_CCC -DHAS_FDESCENTTBL -DFREEBSDV=13000 -DHASFDESCFS=2 -DHASPSEUDOFS -DHASNULLFS -DHASIPv6 -DHASUTMPX -DHAS_STRFTIME -DLSOF_VSTR="13.0-CURRENT" -I/usr/src/sys -g -c ckkv.c -o ckkv.o
In file included from /usr/src/sys/sys/file.h:44,
                 from ../dlsof.h:406,
                 from ../lsof.h:250,
                 from ckkv.c:42:
/usr/src/sys/sys/refcount.h: In function 'refcount_acquire_checked':
/usr/src/sys/sys/refcount.h:65: error: 'false' undeclared (first use in this function)
/usr/src/sys/sys/refcount.h:65: error: (Each undeclared identifier is reported only once
/usr/src/sys/sys/refcount.h:65: error: for each function it appears in.)
/usr/src/sys/sys/refcount.h:68: error: 'true' undeclared (first use in this function)
/usr/src/sys/sys/refcount.h: In function 'refcount_release':
/usr/src/sys/sys/refcount.h:81: error: 'false' undeclared (first use in this function)
/usr/src/sys/sys/refcount.h:90: error: 'true' undeclared (first use in this function)
/usr/src/sys/sys/refcount.h: In function 'refcount_acquire_if_not_zero':
/usr/src/sys/sys/refcount.h:106: error: 'false' undeclared (first use in this function)
/usr/src/sys/sys/refcount.h:108: error: 'true' undeclared (first use in this function)
/usr/src/sys/sys/refcount.h: In function 'refcount_release_if_not_last':
/usr/src/sys/sys/refcount.h:121: error: 'false' undeclared (first use in this function)
/usr/src/sys/sys/refcount.h:123: error: 'true' undeclared (first use in this function)
*** Error code 1

Stop.
make[1]: stopped in /root/testing/lsof/lib
*** Error code 1

Hi, so checking out https://github.com/lrosenman/lsof/tree/fix-refcount gets me back to this:

cc -o lsof  -pipe -DNEEDS_BOOL_TYPEDEF -DHASTASKS -DHAS_PAUSE_SBT -DHAS_DUP2 -DHAS_CLOSEFROM -DHASEFFNLINK=i_effnlink -DHASF_VNODE -DHAS_FILEDESCENT -DHAS_TMPFS -DHASWCTYPE_H -DHASSBSTATE -DHAS_KVM_VNODE -DHAS_UFS1_2 -DHAS_NO_IDEV -DHAS_VM_MEMATTR_T -DNEEDS_DEVICE_T -DHAS_CDEV2PRIV -DHAS_NO_SI_UDEV -DHAS_SYS_SX_H -DHASFUSEFS -DHAS_ZFS -DHAS_V_LOCKF -DHAS_LOCKF_ENTRY -DHAS_NO_6PORT -DHAS_NO_6PPCB -DNEEDS_BOOLEAN_T -DHAS_SB_CCC -DHAS_FDESCENTTBL -DFREEBSDV=13000 -DHASFDESCFS=2 -DHASPSEUDOFS -DHASNULLFS -DHASIPv6 -DHASUTMPX -DHAS_STRFTIME -DLSOF_VSTR=\"13.0-CURRENT\" -I/usr/src/sys -O2 dmnt.o dnode.o dnode1.o dnode2.o dproc.o dsock.o dstore.o  arg.o main.o misc.o node.o print.o proc.o store.o usage.o  util.o -L./lib -llsof  -lkvm
dnode2.o:(.text+0x170): undefined reference to `abd_zero_off'
dnode2.o:(.text+0x1a8): undefined reference to `abd_cmp_buf_off'
dnode2.o:(.text+0x1e0): undefined reference to `abd_copy_to_buf_off'
dnode2.o:(.text+0x218): undefined reference to `abd_copy_from_buf_off'
dnode2.o:(.text+0x254): undefined reference to `abd_copy_off'
*** Error code 1

running: objdump -d -S -l dnode2.o

dnode2.o:     file format elf64-powerpc-freebsd


  88:   38 a0 01 10     li      r5,272
  8c:   48 00 00 01     bl      8c <.readzfsnode+0x48>
  90:   4f ff fb 82     crmove  4*cr7+so,4*cr7+so
  94:   2f a3 00 00     cmpdi   cr7,r3,0
  98:   41 9e 00 20     beq-    cr7,b8 <.readzfsnode+0x74>
  9c:   38 21 11 30     addi    r1,r1,4400
  a0:   e8 62 00 08     ld      r3,8(r2)
  a4:   e8 01 00 10     ld      r0,16(r1)
  a8:   eb c1 ff f0     ld      r30,-16(r1)
  ac:   7c 08 03 a6     mtlr    r0
  b0:   eb e1 ff f8     ld      r31,-8(r1)
  b4:   4e 80 00 20     blr
  b8:   2f bf 00 00     cmpdi   cr7,r31,0
  bc:   e8 01 00 80     ld      r0,128(r1)
  c0:   e9 61 01 10     ld      r11,272(r1)
  c4:   39 20 00 01     li      r9,1
  c8:   f8 1e 00 00     std     r0,0(r30)
  cc:   99 3e 00 28     stb     r9,40(r30)
  d0:   f9 7e 00 10     std     r11,16(r30)
  d4:   41 9e 00 10     beq-    cr7,e4 <.readzfsnode+0xa0>
  d8:   e8 61 00 70     ld      r3,112(r1)
  dc:   2f a3 00 00     cmpdi   cr7,r3,0
  e0:   40 9e 00 

000000000000015c <.abd_zero>:
 15c:   7c 08 02 a6     mflr    r0
 160:   7c 85 23 78     mr      r5,r4
 164:   38 80 00 00     li      r4,0
 168:   f8 01 00 10     std     r0,16(r1)
 16c:   f8 21 ff 91     stdu    r1,-112(r1)
 170:   48 00 00 01     bl      170 <.abd_zero+0x14>
 174:   4f ff fb 82     crmove  4*cr7+so,4*cr7+so
 178:   38 21 00 70     addi    r1,r1,112
 17c:   e8 01 00 10     ld      r0,16(r1)
 180:   7c 08 03 a6     mtlr    r0
 184:   4e 80 00 20     blr
 188:   00 00 00 00     .long 0x0
 18c:   00 00 00 01     .long 0x1
 190:   80 00 00 00     lwz     r0,0(0)

0000000000000194 <.abd_cmp_buf>:
 194:   7c 08 02 a6     mflr    r0
 198:   7c a6 2b 78     mr      r6,r5
 19c:   38 a0 00 00     li      r5,0
 1a0:   f8 01 00 10     std     r0,16(r1)
 1a4:
 20c:   38 a0 00 00     li      r5,0
 210:   f8 01 00 10     std     r0,16(r1)
 214:   f8 21 ff 91     stdu    r1,-112(r1)
 218:   48 00 00 01     bl      218 <.abd_copy_from_buf+0x14>
 21c:   4f ff fb 82     crmove  4*cr7+so,4*cr7+so
 220:   38 21 00 70     addi    r1,r1,112
 224:   e8 01 00 10     ld      r0,16(r1)
 228:   7c 08 03 a6     mtlr    r0
 22c:   4e 80 00 20     blr
 230:   00 00 00 00     .long 0x0
 234:   00 00 00 01     .long 0x1
 238:   80 00 00 00     lwz     r0,0(0)

000000000000023c <.abd_copy>:
 23c:   7c 08 02 a6     mflr    r0
 240:   7c a7 2b 78     mr      r7,r5
 244:   38 c0 00 00     li      r6,0
 248:   f8 01 00 10     std     r0,16(r1)
 24c:   38 a0 00 00     li      r5,0
 250:   f8 21 ff 91     stdu    r1,-112(r1)
 254:   48 00 00 01     bl      254 <.abd_copy+0x18>
 258:   4f ff fb 82     crmove  4*cr7+so,4*cr7+so
 25c:   38 21 00 70     addi    r1,r1,112
 260:   e8 01 00 10     ld      r0,16(r1)
 264:   7c 08 03 a6     mtlr    r0
 268:   4e 80 00 20     blr
 26c:   00 00 00 00     .long 0x0
 270:   00 00 00 01     .long 0x1
 274:   80 00 00 00     lwz     r0,0(0)

--- a/00CREDITS
+++ b/00CREDITS
@@ -544,7 +544,7 @@ provided test systems where I was able to do development work.
 	Ivan Melnikov and
 	@jolmg  (github account)
 	yasu@utahime.org (email)
-	jamie@catflap.org
+	jamie@catflap.org (email)
 
 If I have omitted a contributor's name, the fault is wholly mine,
 and I apologize for the error.

--- a/00CREDITS
+++ b/00CREDITS
@@ -348,6 +348,7 @@ provided test systems where I was able to do development work.
 	Janardhan Molumuri
 	Nasser Momtaheni
 	Laurent Montaron
+	Doug Moore
 	Phillip Moore
 	Dmitry Morozovsky
 	John Paul Morrison
@@ -426,6 +427,7 @@ provided test systems where I was able to do development work.
 	Peter Schiffer
 	Michael Schmitz
 	Larry Schwimmer
+	Cy Schubert
 	Hendrik G. Seliger
 	Igor V. Semenyuk
 	Jonathan Sergent
@@ -541,6 +543,8 @@ provided test systems where I was able to do development work.
 	@eranik (github account)
 	Ivan Melnikov and
 	@jolmg  (github account)
+	yasu@utahime.org (email)
+	jamie@catflap.org
 
 If I have omitted a contributor's name, the fault is wholly mine,
 and I apologize for the error.

--- a/00DIST
+++ b/00DIST
@@ -5041,6 +5041,9 @@ July 14, 2018
 		Don't display command usage even when a file (or directory) listed
 		in command line doesn't exist.
 		This closes the github issue #90 reported by @rowlap.
+		
+		[FreeBSD] merge all the FreeBSD specific fixes from the FreeBSD sysutils/lsof port
+
 
 Masatake YAMATO <yamato@redhat.com>, a member of the lsof-org team at GitHub
 May 8, 2019

--- a/dialects/freebsd/dnode.c
+++ b/dialects/freebsd/dnode.c
@@ -57,9 +57,7 @@ _PROTOTYPE(static int lkup_dev_tty,(dev_t *dr, INODETYPE *ir));
 
 
 #if	defined(HAS_TMPFS)
-#define	_KERNEL
 #include <fs/tmpfs/tmpfs.h>
-#undef	_KERNEL
 #endif	/* defined(HAS_TMPFS) */
 
 _PROTOTYPE(static void get_lock_state,(KA_T f));
@@ -561,7 +559,7 @@ process_overlaid_node:
  * Get the pseudo vnode tag type for FreeBSD >= 5.
  */
 	vtag = VT_UNKNOWN;
-	if (v->v_tag && !kread((KA_T)v->v_tag, (char *)&vtbuf, sizeof(vtbuf)))
+	if (!kread((KA_T)v->v_lock.lock_object.lo_name, (char *)&vtbuf, sizeof(vtbuf)))
 	{
 	    vtbuf[sizeof(vtbuf) - 1] = '\0';
 	    vtbp = vtbuf;

--- a/dialects/freebsd/dnode2.c
+++ b/dialects/freebsd/dnode2.c
@@ -48,7 +48,13 @@ static char copyright[] =
  *
  * Note: clang's complaint about VOP_FSYNC can't be avoided.
  */
-#define	VOP_UNLOCK(vp, f)	((void)0)
+#include <sys/param.h>
+#if __FreeBSD_version >= 1300074
+#define VOP_UNLOCK_FLAGS(vp, f) ((void)0)
+#define VOP_UNLOCK(vp)  ((void)0)
+#else
+#define VOP_UNLOCK(vp, f)       ((void)0)
+#endif
 # endif	/* defined(__clang__) */
 
 #define	KLD_MODULE		/* for ARM: prevent "ARM_NARCH is 0 " error */

--- a/dialects/freebsd/dproc.c
+++ b/dialects/freebsd/dproc.c
@@ -36,6 +36,14 @@ static char copyright[] =
 
 #include "lsof.h"
 
+/*
+ * This is not an exact version but it should not matter. At worst there
+ * is a small version window where this lsof does not compile on older
+ * -CURRENT.
+ */
+#if __FreeBSD_version >= 1300081
+#define HAS_PWD
+#endif
 
 _PROTOTYPE(static void enter_vn_text,(KA_T va, int *n));
 _PROTOTYPE(static void get_kernel_access,(void));
@@ -130,6 +138,15 @@ gather_proc_info()
 	struct fdescenttbl fdt;
 	KA_T fa;
 #endif	/* defined(HAS_FDESCENTTBL) */
+
+#if	defined(HAS_PWD)
+	struct pwd pwd;
+	KA_T pwd_addr;
+#endif	/* defined(HAS_FDESCENTTBL) */
+
+	struct vnode *cdir;
+	struct vnode *rdir;
+	struct vnode *jdir;
 
 	static ofb_t *ofb = NULL;
 	static int ofbb = 0;
@@ -311,6 +328,22 @@ gather_proc_info()
 		continue;
 #endif	/* defined(HAS_FDESCENTTBL) */
 
+#if	defined(HAS_PWD)
+	    cdir = rdir = jdir = NULL;
+	    pwd_addr = (KA_T)FILEDESC_KVM_LOAD_PWD(&fd);
+	    if (pwd_addr != 0) {
+		    if (!kread(pwd_addr, (char *)&pwd, sizeof(pwd))) {
+			    cdir = pwd.pwd_cdir;
+			    rdir = pwd.pwd_rdir;
+			    jdir = pwd.pwd_jdir;
+		    }
+	    }
+#else
+	    cdir = fd.fd_cdir;
+	    rdir = fd.fd_rdir;
+	    jdir = fd.fd_jdir;
+#endif
+
 	/*
 	 * Allocate a local process structure.
 	 */
@@ -346,20 +379,20 @@ gather_proc_info()
 	/*
 	 * Save current working directory information.
 	 */
-	    if (!ckscko && fd.fd_cdir) {
+	    if (!ckscko && cdir) {
 		alloc_lfile(CWD, -1);
 		Cfp = (struct file *)NULL;
-		process_node((KA_T)fd.fd_cdir);
+		process_node((KA_T)cdir);
 		if (Lf->sf)
 		    link_lfile();
 	    }
 	/*
 	 * Save root directory information.
 	 */
-	    if (!ckscko && fd.fd_rdir) {
+	    if (!ckscko && rdir) {
 		alloc_lfile(RTD, -1);
 		Cfp = (struct file *)NULL;
-		process_node((KA_T)fd.fd_rdir);
+		process_node((KA_T)rdir);
 		if (Lf->sf)
 		    link_lfile();
 	    }
@@ -368,10 +401,10 @@ gather_proc_info()
 	/*
 	 * Save jail directory information.
 	 */
-	    if (!ckscko && fd.fd_jdir) {
+	    if (!ckscko && jdir) {
 		alloc_lfile("jld", -1);
 		Cfp = (struct file *)NULL;
-		process_node((KA_T)fd.fd_jdir);
+		process_node((KA_T)jdir);
 		if (Lf->sf)
 		    link_lfile();
 	    }
@@ -654,6 +687,28 @@ kread(addr, buf, len)
 	return((br == len) ? 0 : 1);
 }
 
+static int
+vm_map_reader(void *token, vm_map_entry_t addr, vm_map_entry_t dest)
+{
+	return (kread((KA_T)addr, (char *)dest, sizeof(*dest)));
+}
+
+#if	__FreeBSD_version < 1300060
+typedef int vm_map_entry_reader(void *token, vm_map_entry_t addr,
+    vm_map_entry_t dest);
+
+static inline vm_map_entry_t
+vm_map_entry_read_succ(void *token, struct vm_map_entry *const clone,
+     vm_map_entry_reader reader)
+{
+	vm_map_entry_t next;
+
+	next = clone->next;
+	if (!reader(token, next, clone))
+		return (NULL);
+	return (next);
+}
+#endif	/*  __FreeBSD_version < 1300060 */
 
 /*
  * process_text() - process text information
@@ -681,20 +736,15 @@ process_text(vm)
 /*
  * Read the vm_map structure.  Search its vm_map_entry structure list.
  */
+	vmme = vmsp.vm_map.header;
+	e = &vmme;
 	for (i = 0; i < vmsp.vm_map.nentries; i++) {
 
 	/*
 	 * Read the next vm_map_entry.
 	 */
-	    if (i == 0)
-		e = &vmsp.vm_map.header;
-	    else {
-		if (!(ka = (KA_T)e->next))
-		    return;
-		e = &vmme;
-		if (kread(ka, (char *)e, sizeof(vmme)))
-		    return;
-	    }
+	    if (!vm_map_entry_read_succ(NULL, e, vm_map_reader))
+		return;
 
 #if	defined(MAP_ENTRY_IS_A_MAP)
 	    if (e->eflags & (MAP_ENTRY_IS_A_MAP|MAP_ENTRY_IS_SUB_MAP))

--- a/dialects/freebsd/dsock.c
+++ b/dialects/freebsd/dsock.c
@@ -181,7 +181,7 @@ process_socket(sa)
 #endif	/* FREEBSDV<4050 */
 
 #if	defined(HASIPv6) && !defined(HASINRIAIPv6)
-	struct in6pcb in6p;
+	struct inpcb in6p;
 #endif	/* defined(HASIPv6) && !defined(HASINRIAIPv6) */
 
 	(void) snpf(Lf->type, sizeof(Lf->type), "sock");

